include/cuco/detail/hyperloglog/hyperloglog.cuh

```diff
@@ -0,0 +1,196 @@
+  {
+    this->merge_async(other.ref(), stream);
+  }
+
+  template <cuda::thread_scope OtherScope, class OtherAllocator>
+  void merge(hyperloglog<T, Precision, OtherScope, Hash, OtherAllocator> const& other,
+             cuco::cuda_stream_ref stream = {})
+  {
+    this->merge_async(other, stream);
+    stream.synchronize();
+  }
+
+  template <cuda::thread_scope OtherScope>
+  void merge_async(ref_type<OtherScope> const& other, cuco::cuda_stream_ref stream = {}) noexcept
+  {
+    auto constexpr block_size = 1024;
+    cuco::hyperloglog_ns::detail::merge<<<1, block_size, 0, stream>>>(other, this->ref());
+  }
+
+  template <cuda::thread_scope OtherScope>
+  void merge(ref_type<OtherScope> const& other, cuco::cuda_stream_ref stream = {})
+  {
+    this->merge_async(other, stream);
+    stream.synchronize();
+  }
+
+  [[nodiscard]] std::size_t estimate(cuco::cuda_stream_ref stream) const
+  {
+    // TODO remove test code
+    // std::size_t* result;
+    // cudaMallocHost(&result, sizeof(std::size_t));
+
+    // int grid_size  = 0;
+    // int block_size = 0;
+    // // TODO check cuda error?
+    // cudaOccupancyMaxPotentialBlockSize(
+    //   &grid_size, &block_size, &cuco::hyperloglog_ns::detail::estimate<ref_type<>>);
+
+    // cuco::hyperloglog_ns::detail::estimate<<<grid_size, block_size, 0, stream>>>(
+    //   result, this->ref());
+    // stream.synchronize();
+
+    // return *result;
+
+    // TODO this function currently copies the registers to the host and then finalizes the result;
+    // move computation to device? Edit: host computation is faster -.-
+    storage_type registers;
+    // TODO check if storage is host accessible
+    CUCO_CUDA_TRY(cudaMemcpyAsync(
+      &registers, this->storage_.get(), sizeof(storage_type), cudaMemcpyDeviceToHost, stream));
+    stream.synchronize();
+
+    using fp_type = typename ref_type<>::fp_type;
+    fp_type sum   = 0;
+    int zeroes    = 0;
+    // geometric mean computation + count registers with 0s
+    for (std::size_t i = 0; i < registers.size(); ++i) {
+      auto const reg = registers[i];
+      sum += fp_type{1} / static_cast<fp_type>(1 << reg);
+      zeroes += reg == 0;
+    }
+
+    // pass intermediate result to finalizer for bias correction, etc.
+    return cuco::hyperloglog_ns::detail::finalizer<Precision>::finalize(sum, zeroes);
+  }
+
+  [[nodiscard]] ref_type<> ref() const noexcept
+  {
+    return ref_type<>{*(this->storage_.get()), {}, this->hash_};
+  }
+
+ private:
+  struct storage_deleter {
+    using pointer = typename storage_allocator_type::value_type*;
+
+    storage_deleter(storage_allocator_type& a) : allocator{a} {}
+
+    storage_deleter(storage_deleter const&) = default;
+
+    void operator()(pointer ptr) { allocator.deallocate(ptr, 1); }
+
+    storage_allocator_type& allocator;
+  };
+
+  Hash hash_;
+  storage_allocator_type storage_allocator_;
+  storage_deleter storage_deleter_;
+  std::unique_ptr<storage_type, storage_deleter> storage_;
```

Comment:
So this storage is basically a `aow_storage` with a static extent 1 and the window size is `precision`.

Reply:
Basically yes. Atleast for the dense version of the HLL storage class. The paper also proposes a sparse version which is basically a small hash map.

Reply:
I added some more functionality to the storage class to have a better separation in responsibilities: [cea2afb](https://github.com/NVIDIA/cuCollections/pull/429/commits/cea2afb36f1942f925501041463adda95348ae0f)